### ISSUE TITLE
import in referenced typescript project fail to load

### ISSUE THREAD

**Issue by steabert** [esbuild-kit/tsx#96]
### Bug description

When working with composite projects, our tests failed because it seems that `tsx` is not taking the reference project's baseUrl or path into consideration.

### Reproduction

Test project:

https://github.com/steabert/esbuild-kit-tsx-composite.git

To reproduce the issue:

```
yarn install
yarn fail
```

Running `yarn bundle` and `yarn types` shows the result with e.g. `esbuild` or `tsc`.

It seems this does not work with `ts-node` either, but I don't know how they handle composite typescript projects.

### Environment

```shell
<!--
  Run and paste the output of:
  
  npx envinfo --system --npmPackages tsx --binaries
  ```
-->

  System:
    OS: Linux 5.4 Ubuntu 20.04.4 LTS (Focal Fossa)
    CPU: (8) x64 Intel(R) Core(TM) i5-8250U CPU @ 1.60GHz
    Memory: 23.58 GB / 31.21 GB
    Container: Yes
    Shell: 5.0.17 - /bin/bash
  Binaries:
    Node: 16.14.2 - ~/.node/bin/node
    Yarn: 3.2.3 - ~/.npm-global/bin/yarn
    npm: 8.5.0 - ~/.node/bin/npm
  npmPackages:
    tsx: 3.9.0 => 3.9.0
```


### Can you contribute a fix?

- [ ] I’m interested in opening a pull request for this issue.

**Comment by steabert** [comment]
Also, looks similar to #76 maybe?

**Comment by IlyaSemenov** [comment]
Yes, this is the same problem as reported in #76. unlike esbuild, tsx ignores tsconfig.json in referenced packages.

**Comment by privatenumber** [comment]
How does esbuild handle custom `tsconfig.json` paths via the `tsconfig` CLI flag for multiple projects?

e.g. If project A uses `tsconfig.a.json` and its dependency project B uses `tsconfig.b.json`, how is it configured?

**Comment by steabert** [comment]
I don't need to specify any configuration, esbuild seems to find the correct tsconfig associated with an imported module. However, I think it only does that when bundling. I forgot where I read that, I think it was in the comments on one of the esbuild issues, I'll try to find it again.

**Comment by steabert** [comment]
My own (naive?) way would be to walk up the directory tree of an imported file if there is no associated tsconfig, and if a new `tsconfig.json` file is found, check its include files and cache it for later use. That way, each included file would have an associated tsconfig file.

**Comment by privatenumber** [comment]
My question is more about API design than your specific use-case.

Understanding how esbuild's API handles custom `tsconfig.json` paths across multiple projects would help.

**Comment by steabert** [comment]
The esbuild author replied in my issue, but that seems to be also about how esbuild itself finds tsconfig files. That doesn't seem to be the info you're after, I'm afraid I haven't dealt with the esbuild API at all.

**Comment by privatenumber** [comment]
To be fair, you asked him a different question from the one I asked you.

I don't have time for this now, but if you're interested in helping, you can investigate by setting up a project with the structure I described [here](https://github.com/esbuild-kit/tsx/issues/96#issuecomment-1253726116) and trying to make it work with esbuild. Maybe @IlyaSemenov can help.

Based on what Evan said, I'm skeptical if esbuild is also following TS behavior correctly (not that tsx currently is either). For example, does it even allow for custom `tsconfig.json` paths? And does it check against `tsconfig.json#files/includes/excludes`?

**Comment by mtqperson** [comment]
We are also running into this problem. I might be able to help but I'm not sure what you mean by


> you can investigate by setting up a project with the structure [...] and trying to make it work with esbuild.

> How does esbuild handle custom `tsconfig.json` paths via the `tsconfig` CLI flag for multiple projects?
> e.g. If project A uses `tsconfig.a.json` and its dependency project B uses `tsconfig.b.json`, how is it configured?

Like @steabert described, with esbuild it already works. Do you want to know whether it still works if we additionally specify tsconfigs manually, even though that's not required?

**Comment by privatenumber** [comment]
I'm skeptical that esbuild would successfully build. My understanding is that it only detects `tsconfig.json`, and not `tsconfig.b.json`.

To prove that this works, please create a repository containing a monorepo with two tsconfigs under custom names and add a build script to demonstrate that esbuild can successfully build it. If you'd like, you can reuse https://github.com/steabert/esbuild-kit-tsx-composite but rename the `tsconfig.json` of the dependency and test if it builds.

And if that doesn't build, please find out what configuration esbuild offers to specify a custom `tsconfig.json` name.

**Comment by privatenumber** [comment]
We should figure out how custom tsconfig paths are handled in nested projects: https://github.com/esbuild-kit/tsx/issues/96#issuecomment-1253726116

**Comment by IlyaSemenov** [comment]
I am not following why the custom configs matter so much? It would suffice if nested projects worked at least with default configs. Supporting custom configs in nested projects could be the Phase 2 implemented by whoever cares about that use case.

Chances are esbuild doesn't support custom configs in nested projects at all, then what? That doesn't prevent it from working with default configs; and neither should it prevent tsx from supporting the same case.

**Comment by privatenumber** [comment]
> I am not following why the custom configs matter so much? It would suffice if nested projects worked at least with default configs.

Why it matters: https://github.com/esbuild-kit/tsx/issues/76#issuecomment-1274526403

tsx has a lot of users and handles many use-cases. Custom tsconfig files are very common so I think the more appropriate question you should be answering is "why should we dismiss those use-cases without even doing an hour of research?".

> Supporting custom configs in nested projects could be the Phase 2 implemented by whoever cares about that use case.

FYI, I don't benefit from this use-case but I'm happy to implement it in the future. With regards to your attitude, I'm wondering why you haven't implemented this yourself? As with anything open-source, you should also feel free to implement it yourself and use your own fork.


> Chances are esbuild doesn't support custom configs in nested projects at all, then what? That doesn't prevent it from working with default configs; and neither should it prevent tsx from supporting the same case.

That's fine too. But no one has provided a definitive answer with research.

When I decide to take this on, I would like to think of the best solution possible and would investigate different use-cases and solutions. Since ya'll are asking how you can help, I provided one of the questions I would research.

----

I'm not sure how this thread will be productive if my answers to your "how can I help?" is met with "why is that important?".

If no one is going to provide insight into potential problems and how other projects handle it, it might be best to lock this thread until I'm ready to tackle it and do the research myself.

**Comment by IlyaSemenov** [comment]
Thank you for the response. I apologize if I sounded rude. I know how open-source works, I authored and support a number of npm and python modules myself, and I am happy to provide users the right direction if they ask (or point them where they're wrong).

My point here was that making fully broken system half-broken would still be a good thing. I gathered that the current implementation blocker was that it wasn't clear how to configure esbuild to use different tsconfig's in different sub-projects. In my thinking that should not be a blocker even if that's not possible.

I did make the initial research back then when you asked the first time and it immediately appeared that the public esbuild API only allowed to customize the root/entrypoint tsconfig: https://esbuild.github.io/api/#tsconfig
I didn't share that finding as it appeared too obvious for me.

In light of that, I was merely asking if you could provide your initial high level thoughts of where the source could be extended to load paths from nested tsconfigs (considering the limitation of not being able to use custom-name tsconfig), outline the frames of the possible required architectural changes. I surely understand it's not always possible to answer that kind of questions, but sometimes it is.

**Comment by jgoux** [comment]
Hello everyone 👋,

I also hit this issue. There is no way to specify custom `tsconfig.<whatever>.json` config files for external packages bundled with esbuild. It only looks for the regular `tsconfig.json`.

As far as I understand, just supporting the regular `tsconfig.json` files would solve my issue. For that, we should not pass an explicit  `--tsconfig` option when calling esbuild, as it will use this single tsconfig file and ignore all the others.

So `paths` defined in `tsconfig.json` can work but in a limited way (every tsconfig files need to be named `tsconfig.json`, including the root one).

Do you think we could add an option to enable this behavior? Maybe `--resolve-tsconfigs` or something like this?

I'm happy to make a PR if you can point me to the relevant files! 😄

**Comment by jgoux** [comment]
If people are stuck on this, I had success with [vite-node](https://www.npmjs.com/package/vite-node) combined with [vite-tsconfig-paths](https://www.npmjs.com/package/vite-tsconfig-paths).

It also plays nice with vitest as it's reusing the same pipeline. 👍

**Comment by jgoux** [comment]
I'm very sorry for this triple post, but I played some more with my TypeScript monorepo and I think I gained interesting insights for this issue.

## Big picture

> tldr; TypeScript project references are the way to hint about which `tsconfig.json` file to apply to parts of the dependency chain

First, the "one true way" of connecting TypeScript projects together is [project references](https://www.typescriptlang.org/docs/handbook/project-references.html). This is the only mechanism that enables TypeScript to split the dependency chain between multiple TypeScript projects and use each project's `tsconfig.json`.

Custom `tsconfig.json` files are handled by this mechanism as well:

```jsonc
{
    "compilerOptions": {
        // The usual
    },
    "references": [
        { "path": "../api/tsconfig.build.json" },
        { "path": "../sdk/tsconfig.whatever.json" },
    ]
}
```

Without references, if you require a TypeScript file from an external package (through either monorepo workspaces or paths alias, I personally recommend avoiding paths alias as you break your package isolation), only the `tsconfig.json` file of the root package will be used, this is why external packages paths alias break, because the root package is not aware of their internals (and it should not be 😄).

## How to have this implemented in tsx

Project references were [discussed in esbuild repository](https://github.com/evanw/esbuild/issues/1250#issuecomment-841704482), and the conclusion was that it's beyond the scope of esbuild, but could be implemented as a plugin!

A plugin exists: [esbuild-plugin-ts-references](https://github.com/smacker/esbuild-plugin-ts-references) but it's not working on paths alias, it's only rewriting the import statements to link directly to the references. But it shows that it's possible with esbuild API to walk through each package's `tsconfig.json` by following references and build some kind of map.

So the solution would be to write an esbuild plugin similar to [esbuild-plugin-ts-references](https://github.com/smacker/esbuild-plugin-ts-references) that can rewrite imports based on the file's local `tsconfig.json` gathered through references analysis. The main blocker is that [`tsx` is using esbuild's Transform API and not the Build API, so it can't consume a plugin](https://github.com/esbuild-kit/tsx/discussions/50#discussioncomment-3027253).

@privatenumber How hard of a requirement is it to use the Transform API vs the Build API? Reading esbuild doc it seems like Transform is just a small subset of Build?

Another solution would be to have `tsx` do the references analysis work outside of esbuild to build the map and then apply the rewrites somehow with the Transform API, but I'm not familiar with it enough to know if it's possible. 😅

**Comment by jgoux** [comment]
I opened PRs to esm-loader and cjs-loader to support project references which should solve this issue.

@steabert @IlyaSemenov If you can give it a try it would be great! 👍

**Comment by pokey** [comment]
I was able to get this working using a custom `exports` field on all my monorepo `package.json` files, and then passing `conditions` to `tsx`:

```json
  "exports": {
    ".": {
      "bundler": "./src/index.ts",
      "default": "./out/index.js"
    }
  }
```

Then

```bash
tsx --conditions=bundler ...
```

**Comment by guillaume-macquat** [comment]
I had a similar issue in a monorepo where path aliases of imported workspaces were not being resolved using their own `tsconfig.json`. I think tsx should be able to detect the config file associated with each module, so I changed this behavior locally and here's my version of `src/utils/tsconfig.ts`:
```ts
import { getTsconfig, createPathsMatcher } from 'get-tsconfig';

const configCache = new Map<string, any>();

export function fileMatcher(filename: string) {
  return getTsconfig(filename, 'tsconfig.json', configCache)?.config;
};

export function tsconfigPathsMatcher(specifier: string, parentPath: string = process.cwd()) {
  const resolvedConfig = getTsconfig(parentPath, 'tsconfig.json', configCache);
  return resolvedConfig
    ? createPathsMatcher(resolvedConfig)!(specifier)
    : [];
};

export const allowJs = false;

export const loadTsconfig = (configPath?: string) => {};
```
The idea is that when `fileMatcher` or `tsconfigPathsMatcher` are called they use a reference file path to resolve the appropriate `tsconfig.json`. Note that I had to add an argument to `tsconfigPathsMatcher` and change its calls. This code suits my needs and I don't know tsx enough to provide a pull request. To continue on this approach one would have to:
- Change the logic of the `allowJs` export to be dynamic.
- Completely remove `loadTsconfig`.
- Continue supporting tsx `tsconfig` option and define a logic to support both strategies.

Another approach could be to change the get-tsconfig package to handle project references. In monorepos we can have a root tsconfig that references every workspace tsconfigs. Given a root tsconfig and a specified file path get-tsconfig could generate a config that represents all the options applied to it. This way we could provide a single config in tsx options.

**Comment by guillaume-macquat** [comment]
Why was my comment marked off-topic when I directly provided a solution to this issue and some of the related issue that were closed in favor of this one?
This issue was created 2 years ago and have been inactive for a year and a half. Rejecting contribution at this point is just ridiculous.
@privatenumber you insist on this project being open-source and yet you disregard contribution that does not satisfy your exclusive scrutiny. This issue is not about esbuild, it's about tsx. If you don't agree that my contribution is the way to go then someone else certainly might.
You mentioned that the maintenance of these threads uses your time. Don't forget that very often these issues save time for users that are still looking for a solution. Your contribution of marking my contribution off-topic is likely to waste someone's time. I think your contribution is off-topic.

**Comment by privatenumber** [comment]
@guillaume-macquat

> Why was my comment marked off-topic when I directly provided a solution to this issue and some of the related issue that were closed in favor of this one?

I understand your frustration, but your comment appeared to be more of a workaround than a step towards a concrete solution for tsx itself. It also didn't indicate any intent or effort to contribute to the codebase, which is a crucial aspect of this project's collaborative nature. For instance, you said:

> This code suits my needs and I don't know tsx enough to provide a pull request.

In the future, I would recommend starting by asking how you can contribute these changes into tsx instead and I'd be happy to guide you.

I encourage you to read the [Contribution Guide](https://github.com/privatenumber/tsx/blob/master/CONTRIBUTING.md#commenting-on-an-existing-issue), which outlines how to contribute effectively and engage in discussions that help move the project forward.

> Don't forget that very often these issues save time for users that are still looking for a solution.

I understand and appreciate that sharing workarounds can sometimes be helpful. However, this repository isn't intended as a forum for one-off solutions without contributing to the core code. The Issues tab is specifically for tracking concrete changes or improvements to tsx, and keeping the focus there ensures better organization and maintainability.

I also want to address the tone of your comment, which came off as confrontational. I value open discussions, but they need to remain respectful and productive. I'd recommend adopting a more collaborative tone to foster a positive conversation in any community.

With that said, I'll be hiding off-topic comments (including mine) and locking this thread, as this issue will be addressed in the upcoming major release.
